rfcs/0132-meson-builds-nix.md

```diff
@@ -0,0 +1,127 @@
+
+In particular, many changes have been introduced that impact compatibility outside Linux and NixOS niches. These issues can hinder development on other platforms, including but not limited to Unix-like systems.
+
+In light of this state of things, propose a novel alternative to the current buildsystem infrastructure.
+
+We expect to accomplish, among other goals,
+
+- better code structuring and documentation;
+- improved cross-platform support, especially in other programming environments, including but not limited to Unix-like operating systems;
+- shorter build times;
+- improved unit testing;
+- an overall improved user experience.
+
+# Detailed design
+[design]: #detailed-design
+
+A carefully crafted set of Meson files should be included in the Nix repository, in order to describe how to deploy the Nix repository, generating all the expected artifacts (command line tools, libraries, configuration files etc.)
+
+This novel build infrastructure should be able to provide at least feature parity with the current  quasi-autotools implementation, albeit with a different user interface.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Currently @p01arst0rm is writing an implementation from scratch.
+
+Some typical expected interactions are:
+
+- Configuring: `meson setup build_dir`
+- Building: `ninja -C build_dir` (or the backend-agnostic `meson -C build_dir compile`)
+- Installing: `ninja -C build_dir install` (or the backend-agnostic `meson -C build_dir install`)
+
+Usually, commandline options assume the format `-Dname=value`. This small change in user interface when compared to the typical `--name=value` is mostly harmless.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Some possible drawbacks:
+
+- A new build system would require changes on the code
+  + On the other hand, such changes are likely to improve the code base.
+  
+- A new build system requires the developers become familiarized with it
+  - Specially when this build system uses its own description language.
+  + However, Meson is well documented, and its Python-esque language is easy to grasp.
+
+- A new build system indirectly brings its own dependencies to the Nix project
+  - In particular, the reference implementation of Meson is written in Python.
+  - Further, this reference implementation generates script files meant to be consumed by Ninja.
+  - Ninja is a tool written in C++ that acts like a `make` replacement.
+  - This particular setting brings concerns about complexifying the bootstrap route.
+  + Given that Nix is currently written in C++, we can assume a C++ compiler as part of such a bootstrap route.
+  + There are full-featured alternative tools that replace Meson and Ninja. 
+    + Namely, Muon and Samurai are implementations of Meson and Ninja that require only a C compiler and a set of POSIX standard tools.
```

Comment:
I would definitely keep this drawback in mind and explore alternatives that don't expand the bootstrap closure.

Maybe you could also mention what the current bootstrap closure is?

Reply:
@asymmetric please read the prior PR to read discussions; meson does not increase bootstrap dependencies, it dramatically decreases them. available tools are provided that only require a working c compiler; c --> muon --> samurai --> nix. the benefit & end goal of this is removing gnu userland tools as a requirement of building, dramatically decreasing the amount of tools needed for bootstrapping.

Reply:
> please read the prior PR to read discussions

You mean https://github.com/NixOS/nix/pull/3160?

> meson does not increase bootstrap dependencies

Didn't you say in the RFC that Meson is in Python? What I wanted to say with my comment is to emphasize the idea that we shouldn't increase the bootstrap closure, and so I approve that you mentioned it and provided alternatives that don't increase it.

But AFAIU Python is not part of the current Nix bootstrap closure, so if we use straight Meson, rather than muon and/or Ninja, we would indeed increase it. Is that correct?

If that's the case, why wouldn't we choose muon directly?

Reply:
Who said that muon isn't "directly" chosen?

Maybe the confusion here is that the RFC is to use "Meson" the buildsystem description DSL, but not specifically to bootstrap Nix with "Meson" the reference implementation of "Meson".

What's the goal of this review comment? To rename this RFC to

```
[RFC 0132] Muon Builds Nix
```

???

Reply:
> Maybe the confusion here is that the RFC is to use "Meson" the buildsystem description DSL, but not specifically to bootstrap Nix with "Meson" the reference implementation of "Meson".

That was indeed the thing that confused me

Reply:
maybe we need a line in the RFC that says "meson the dsl language != meson.py the tool"

Reply:
> Maybe you could also mention what the current bootstrap closure is?

As far as I know, M4 (the somewhat obscure macro language) and Bash, and whatever autoconf brings.

Reply:
some required build programs include 

- flex
- bison
- sed
- tar
- dot
- jq
- bash
- tar
- patch
- gcc (last time i attempted, clang compilation is very sketch)